ush/set_cycle_dates.py

```diff
@@ -6,59 +6,43 @@
 
 
-def set_cycle_dates(date_start, date_end, cycle_hrs, incr_cycl_freq):
-    """This file defines a function that, given the starting date (date_start,
-    in the form YYYYMMDD), the ending date (date_end, in the form YYYYMMDD),
-    and an array containing the cycle hours for each day (whose elements
-    have the form HH), returns an array of cycle date-hours whose elements
-    have the form YYYYMMDD.  Here, YYYY is a four-digit year, MM is a two-
-    digit month, DD is a two-digit day of the month, and HH is a two-digit
-    hour of the day.
+def set_cycle_dates(date_start, date_end, incr_cycl_freq):
+    """This file defines a function that, given the start and end dates
+    as date time objects, and a cycling frequency, returns an array of
+    cycle date-hours whose elements have the form YYYYMMDDHH.  Here,
+    YYYY is a four-digit year, MM is a two- digit month, DD is a
+    two-digit day of the month, and HH is a two-digit hour of the day.
 
     Args:
-        date_start: start date
-        date_end: end date
-        cycle_hrs: [ HH0, HH1, ...]
-        incr_cycl_freq: cycle frequency increment in hours
+        date_start: start date, datetime object
+        date_end: end date, datetime object
+        incr_cycl_freq: cycle frequency increment in hours, an int
     Returns:
         A list of dates in a format YYYYMMDDHH
     """
 
     print_input_args(locals())
 
-    # calculate date increment
-    if incr_cycl_freq <= 24:
-        incr_days = 1
-    else:
-        incr_days = incr_cycl_freq // 24
-        if incr_cycl_freq % 24 != 0:
-            print_err_msg_exit(
-                f'''
-                INCR_CYCL_FREQ is not divided by 24:
-                  INCR_CYCL_FREQ = \"{incr_cycl_freq}\"'''
-            )
+    freq_delta = timedelta(hours=incr_cycl_freq)
 
-    # iterate over days and cycles
+    # iterate over cycles
     all_cdates = []
-    d = date_start
-    while d <= date_end:
-        for c in cycle_hrs:
-            dc = d + timedelta(hours=c)
-            v = datetime.strftime(dc, "%Y%m%d%H")
-            all_cdates.append(v)
-        d += timedelta(days=incr_days)
-
+    cdate = date_start
+    while cdate <= date_end:
+        cyc = datetime.strftime(cdate, "%Y%m%d%H")
+        all_cdates.append(cyc)
+        cdate += freq_delta
     return all_cdates
 
 
 class Testing(unittest.TestCase):
     def test_set_cycle_dates(self):
         cdates = set_cycle_dates(
-            date_start=datetime(2022, 1, 1),
-            date_end=datetime(2022, 1, 4),
-            cycle_hrs=[6, 12],
-            incr_cycl_freq=48,
+            date_start=datetime(2022, 1, 1, 6),
+            date_end=datetime(2022, 1, 2, 12),
+            incr_cycl_freq=6,
         )
         self.assertEqual(
-            cdates, ["2022010106", "2022010112", "2022010306", "2022010312"]
+            cdates, ["2022010106", "2022010112", "2022010118",
+                "2022010200", "2022010206", "2022010212"]
```

Comment:
I changed this test to reflect the new behavior. I do not feel that this test necessarily reflects a standard use case for how folks choose to cycle their experiments. The original tested that cycles would be set for 6 and 12 UTC 2 days apart. It is much more standard for cycling capabilities to be set at regular intervals.

I am currently working on introducing changes to XML creation that should allow us to set up any number of cycledefs using any of the native Rocoto cycledef options in a more straightforward way.

I hope that the change in behavior to this script will be tolerated in the meantime.

Reply:
The current workflow allows a user to choose any combination of forecast cycles between 00 and 23Z to be repeated between DATE_FIRST_CYCL and DATE_LAST_CYCL to provide as much flexibility to the community as possible.  While it may not be a use case in operational meteorology, it's very possible that other users in the community could find it useful for their purposes, particularly for those in academia or researchers working on model or physics improvements.  For example, a developer may wish to only run during nocturnal cycle hours to test physics innovations intended to improve results during the night, and have no need to run cycles during the day.  Running only during nocturnal hours over many days allows that user to create a large sample size to produce statistically significant results, while limiting cycles (i.e., disk space and compute time) that they don't need.  Will the changes to XML creation you mention allow a user to choose any given set of forecast cycles such as how the functionality works now?  Would those changes be accompanied by a user-defined variable such as "CYCL_HRS" where a set of cycles can be defined?  If that's the case and the "tolerated change in behavior" is just this limitation for the time being, I think that's fine, but just wanted to make sure.  @gsketefian, @mkavulich, @BenjaminBlake-NOAA, @mark-a-potts, @danielabdi-noaa, @MichaelLueken, thoughts?

Reply:
The new XML generation would not be setting just one start, stop, end that are defined by variables like we currently have. Instead, a user would set any number of cycledefs supported by Rocoto in their yaml config. It might look something like this. Here, I'm recreating the same cycling Daniel originally had in the test case. 

```
  cycledefs:
    at_start:
      dates: !startstopfreq ['DATE_FIRST_CYCL', 'DATE_FIRST_CYCL', 'CYCL_FREQ']
    forecast6z:
      dates: !startstopfreq [2022010106, 2022010406, 48:00:00]
    forecast12z:
      dates: !startstopfreq [2022010112, 2022010412, 48:00:00]
    
```

Reply:
@christinaholtNOAA do you think your future XML creation changes will be ready before the code slush deadline? If so I think @JeffBeck-NOAA's point is moot since the capability will be re-introduced (just in a different way)

Reply:
@christinaholtNOAA, @mkavulich, OK!  So if a user wanted to run, say, 00, 01, 02, 03, and 04Z every day for two weeks, that configuration could be created with five different "dates:" entries for cycledefs?  If that functionality comes in time for v2.1, that would be great.